---
title: "ROS Gene Count Workflow - Get Genome Sequences of ROS BLAST Hits"
author: "Brian Beardsall"
date: "07/05/2020"
output:
  html_document:
    df_print: paged
    toc: true
---

### BASH

Run the following bash script to parse a folder of FASTA files, determine if AA or DNA, make a BLAST database for each file (prot or nucl), and BLAST a FASTA file of probe sequences (either tblastn or blastp) against each database. Save results to csv.

```{}
cd Bash
bash ParseAndBlast.bash
```

### Setup 

#### Install Packages 

Note that the `Biostrings`package from Bioconductor is required, which can be installed through `BiocManager`.
Install Bioconductor packages, if not already installed.
```{r BioStringsInstall}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}
    
if (!requireNamespace("Biostrings", quietly = TRUE)) {
  install.packages("Biostrings")
}
```

#### Load libraries.
```{r setup, message=FALSE,warning=FALSE}
library(tidyverse, quietly = TRUE)
library(Biostrings, quietly = TRUE)
```

#### Assign filepaths.
```{r assign filepaths}
InDataPath <- file.path("..", "DataIn")
ProcessDataPath <- file.path("..", "ProcessData")

BlastResultsPath <- file.path(ProcessDataPath, "BlastResults")
BlastDatabasesPath <- file.path(ProcessDataPath, "BlastDatabases")
GenomesPath <- file.path(InDataPath,"Genomes")
BlastHitSeqsPath <- file.path(ProcessDataPath,"BlastHitSeqs")
```

### Read BLAST Results

Function to read in a blast results csv, and keep only the unique sequence IDs (since multiple probes can match the same sequence).
```{r}
read_blast_csv_unique_results <- function(InFilepath, ColNames) {
  read_csv(file = InFilepath, col_names = ColNames) %>%
    # Add cols for filepath, name, genome/transcriptome name, and blast type.
    mutate(Filepath = InFilepath,
           FileName = basename(InFilepath),
           OmeName = separate_blast_result_file_name(InFilepath)[["OmeName"]],
           BlastType = separate_blast_result_file_name(InFilepath)[["BlastType"]]) %>%
    # keep only unique sequences
    distinct_at(vars(sseqid), .keep_all = TRUE)
}
```

Helper function for read_blast_csv_unique_results to extract the BlastType and OmeName from a blast result filepath.
```{r}
separate_blast_result_file_name <- function(InFilePath) {
  # get only the filename from the path
  FileName <- basename(InFilePath)
  # get rid of the file extension
  RemovedEnding <- unlist(str_split(FileName,"\\."))[1] 
  # split at the 
  SplitString <- unlist(str_split(RemovedEnding, "___"))
  names(SplitString) <- c("BlastType", "OmeName")
  return(SplitString)
}
```

Get the blast result filepaths, and genome filepaths.
```{r}
BlastResultFiles <- list.files(path = file.path(BlastResultsPath), full.names = TRUE)
GenomeFiles <- list.files(path = file.path(GenomesPath), full.names = TRUE)
```


The CSV output of BLAST+ doens't have a header, so add one manually.
```{r}
BlastHeaders <- c("qseqid", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "sframe", "qframe")
```

Get the unique sequence IDs from each blast result csv, and combine into one dataframe.
```{r, message=FALSE,warning=FALSE}
UniqueBlastResults <- BlastResultFiles %>%
  map(~read_blast_csv_unique_results(InFilepath = ., ColNames = BlastHeaders)) %>%
  enframe() %>%
  unnest(cols = value)
```


Extract the genome names from the file paths.
```{r}
GenomeNames <- basename(GenomeFiles) %>%
  # split at period
  str_split(string = ., pattern = "\\.") %>%
  # iterate over the list, keeping only the the first element of each vector
  sapply(., "[[", 1)

GenomeNames
```

It's vastly more efficient to work with FASTA format files as AAStringSet or DNAStringSet objects (from `BioStrings`) than character strings. Small sequences would be OK as normal strings, but entire genomes would be unusable.


Function to read a protein FASTA format file, and return an AAStringSet object. 
```{r}
read_ome_as_AAStringSet <- function(filePath){
  # Call AAStringSet constructor
  OmeSeqs <- readBStringSet(filePath)
  
  # Shorten the sequence names to before the space, to match Blast output
   names(OmeSeqs) <- names(OmeSeqs) %>%
    str_split("\\s") %>%
     # Use "[[" operator to extract first element of each list
    sapply("[[", 1)
  
  return(OmeSeqs)
}
```


Iterate through FASTA genome files, saving as AAStringSets in a list, with element names set to genome names.
```{r}
OmeSeqsList <- map(GenomeFiles, read_ome_as_AAStringSet)
names(OmeSeqsList) <- GenomeNames
```

Function to select only sequences that were a Blast hit for ROS probes from each genome, and write to a FASTA file. 

Takes a genome name, and list of XStringSets (could be AAStringSets or DNAStringSets) with a corresponding genome name element.

```{r}
write_csv(UniqueBlastResults, file.path(ProcessDataPath, "AllProbeBlastHits.csv"))
UniqueBlastResults 
```


```{r}
write_blast_hits_fasta <- function(InOmeName, InOmeSeqsList, OutFilePath){
  # get sequence IDs for blast hits from the current genome
  HitGenomeSeqIDs <- UniqueBlastResults %>%
    filter(OmeName == InOmeName) %>%
    pull(sseqid)
  
  # access current genome XStringSet from the list  
  OmeSeqs <- InOmeSeqsList[InOmeName][[1]]
  # subset only the blast hits
  HitOmeSeqs <- OmeSeqs[HitGenomeSeqIDs]
  
  # Add the genome name in front of each sequence ID
  names(HitOmeSeqs) <- paste(InOmeName, names(HitOmeSeqs), sep = "___")
  
  # Write to a FASTA file
  writeXStringSet(HitOmeSeqs, file.path(OutFilePath, 
                                           paste("HITS_", InOmeName, ".fasta", sep = '')))
  
  return(NULL)
}
```

Iterate through the genomes, writing their unique blast result sequences to a FASTA format file.  
Since `write_blast_hits_fasta` returns NULL, a for loop is more fitting than map (but either works).
```{r}
for(i in GenomeNames) {
  write_blast_hits_fasta(i, InOmeSeqsList = OmeSeqsList, OutFilePath = BlastHitSeqsPath)
}
```

```{}
cd Bash
bash ConcatHitSeqs.bash
```

